File: frontend/src/engine/simulation.js
/**
 * SimulationEngine — tick pipeline, world generation, entity management.
 * No threading — designed to be called from a Web Worker.
 */
import { World } from './world.js';
import { Animal, Egg } from './entities.js';
import { SpatialHash } from './spatialHash.js';
import { generateTerrain, computeWaterProximity } from './mapGenerator.js';
import {
  seedInitialPlants,
  processPlants,
  P_APPLE_TREE,
  P_MANGO_TREE,
  P_STRAWBERRY,
  P_BLUEBERRY,
  P_GRASS,
  P_CARROT,
  P_POTATO,
  P_CHILI_PEPPER,
  P_OLIVE_TREE,
  S_ADULT,
} from './flora.js';
import { decideAndAct, giveBirth } from './behaviors.js';
import { BASE_POP_TOTAL } from './animalSpecies.js';
import {
  createBenchmarkCollector,
  resetBenchmarkCollector,
  cloneBenchmarkCollector,
  benchmarkAdd,
} from './benchmarkProfiler.js';

export class SimulationEngine {
  constructor(config) {
    this.config = config;
    this.world = null;
    this.spatialHash = new SpatialHash(16);
    this.profilingEnabled = false;
    this._latestProfile = null;
    this._benchmarkCollector = createBenchmarkCollector();
    this.spatialHash.setBenchmarkCollector(this._benchmarkCollector);
  }

  setProfilingEnabled(enabled) {
    this.profilingEnabled = !!enabled;
  }

  getLatestProfile() {
    return this._latestProfile;
  }

  resetBenchmarkStats() {
    resetBenchmarkCollector(this._benchmarkCollector);
    if (this.world) this.world._benchmarkCollector = this._benchmarkCollector;
    this.spatialHash.setBenchmarkCollector(this._benchmarkCollector);
  }

  getBenchmarkStats() {
    return cloneBenchmarkCollector(this._benchmarkCollector);
  }

  /**
   * Generate a new world with terrain, plants, and animals.
   * Returns the seed used.
   */
  generateWorld() {
    this.world = new World(this.config);
    this.world._benchmarkCollector = this._benchmarkCollector;
    const { terrain, waterProximity, heightmap, seed } = generateTerrain(this.config);
    this.world.terrain = terrain;
    this.world.waterProximity = waterProximity;
    this.world.heightmap = heightmap;

    seedInitialPlants(this.world);
    this._spawnAnimals();
    this.spatialHash.rebuild(this.world.animals);
    return seed;
  }

  /**
   * Reset simulation state (animals, plants, clock, stats) while preserving the terrain map.
   */
  resetSimulation() {
    const w = this.world;
    w._benchmarkCollector = this._benchmarkCollector;
    // Reset plants
    w.plantType.fill(0);
    w.plantStage.fill(0);
    w.plantAge.fill(0);
    w.plantChanges = [];
    w.activePlantTiles.clear();

    // Reset animals
    w.animals = [];
    w.eggs = [];
    w._nextId = 1;
    w.animalGrid.fill(0);

    // Reset clock
    w.clock.tick = 0;

    // Reset stats
    w.statsHistory = [];
    w.resetPlantEvents();

    // Re-seed plants and animals
    seedInitialPlants(w);
    this._spawnAnimals();
    this.spatialHash.rebuild(w.animals);
  }

  _spawnAnimals() {
    const w = this.world;
    const counts = this.config.initial_animal_counts || {};

    for (const [species, count] of Object.entries(counts)) {
      const speciesConfig = this.config.animal_species[species];
      if (!speciesConfig) continue;
      const walkableSet = new Set(speciesConfig.walkable_terrain || [1, 2, 3, 5, 8]);
      // Max initial age: 50% of max_age — gives natural age distribution
      // and prevents synchronized mass die-offs
      const maxInitAge = Math.floor((speciesConfig.max_age || 1000) * 0.5);
      let placed = 0, attempts = 0;
      while (placed < count && attempts < count * 50) {
        const x = Math.floor(Math.random() * w.width);
        const y = Math.floor(Math.random() * w.height);
        if (w.isWalkableFor(x, y, walkableSet) && !w.isTileOccupied(x, y)) {
          const animal = new Animal(w.nextId(), x + 0.5, y + 0.5, species, speciesConfig);
          animal.age = Math.floor(Math.random() * maxInitAge);
          w.animals.push(animal);
          w.placeAnimal(x, y);
          placed++;
        }
        attempts++;
      }
    }
  }

  /**
   * Process one simulation tick (sequential, backward-compatible).
   */
  tick() {
    this.tickFlora();
    this.tickFaunaSequential();
    this.tickEggs();
    this.tickCleanup();
  }

  /**
   * Phase 1: Advance clock and process flora.
   */
  tickFlora() {
    const w = this.world;
    w.plantChanges = [];
    this._tickStart = performance.now();
    this._phases = { plantsMs: 0, behaviorMs: 0, spatialMs: 0, cleanupMs: 0, statsMs: 0 };

    w.clock.advance();

    const plantsStart = performance.now();
    processPlants(w);
    this._phases.plantsMs = performance.now() - plantsStart;
  }

  /**
   * Phase 2 (sequential): Process fauna one-by-one in the current thread.
   */
  tickFaunaSequential() {
    const w = this.world;

    const behaviorStart = performance.now();
    this._deadThisTick = [];
    this._processedAnimals = 0;
    for (const animal of w.animals) {
      if (animal.alive) {
        this._processedAnimals++;
        decideAndAct(animal, w, this.spatialHash);
        if (!animal.alive) {
          this._deadThisTick.push(animal);
        }
      }
    }
    this._phases.behaviorMs = performance.now() - behaviorStart;

    // Spatial hash update for moved/dead animals
    const spatialStart = performance.now();
    for (const animal of w.animals) {
      if (animal.alive) this.spatialHash.update(animal);
    }
    this._phases.spatialMs = performance.now() - spatialStart;
  }

  /**
   * Phase 2 (parallel): Apply merged results from fauna sub-workers.
   * @param {Array} results — array of { deltas, births, plantChanges, deadIds } from each sub-worker.
   */
  applyFaunaResults(results) {
    const w = this.world;
    const animalsById = new Map();
    for (const a of w.animals) animalsById.set(a.id, a);

    // Collect and sort all deltas by id for deterministic merge order
    const allDeltas = [];
    for (const r of results) {
      if (!r.deltas) continue;
      for (const d of r.deltas) allDeltas.push(d);
    }
    allDeltas.sort((a, b) => a.id - b.id);

    // Reset occupancy grid — we'll rebuild it
    w.animalGrid.fill(0);

    // Place unchanged animals first (those not in any delta)
    const deltaIds = new Set(allDeltas.map(d => d.id));
    for (const a of w.animals) {
      if (a.alive && !deltaIds.has(a.id)) {
        w.placeAnimal(a.x, a.y);
      }
    }

    // Apply deltas with movement conflict resolution
    this._deadThisTick = [];
    this._processedAnimals = allDeltas.length;
    for (const delta of allDeltas) {
      const animal = animalsById.get(delta.id);
      if (!animal) continue;

      const moved = delta.x !== animal.x || delta.y !== animal.y;
      let finalX = delta.x, finalY = delta.y;

      // Movement conflict: if new tile is occupied, keep old position
      if (moved && delta.alive && w.isTileOccupied(finalX, finalY)) {
        finalX = animal.x;
        finalY = animal.y;
      }

      // Apply state from sub-worker
      animal.x = finalX;
      animal.y = finalY;
      animal.state = delta.state;
      animal.energy = delta.energy;
      animal.hp = delta.hp;
      animal.hunger = delta.hunger;
      animal.thirst = delta.thirst;
      animal.age = delta.age;
      animal.alive = delta.alive;
      animal.mateCooldown = delta.mateCooldown;
      animal.attackCooldown = delta.attackCooldown;
      animal.path = delta.path || [];
      animal.pathIndex = delta.pathIndex || 0;
      animal._pathTick = delta._pathTick || 0;
      animal._deathTick = delta._deathTick;
      animal.consumed = delta.consumed || false;
      animal.targetX = delta.targetX;
      animal.targetY = delta.targetY;
      animal.pregnant = delta.pregnant || false;
      animal.gestationTimer = delta.gestationTimer || 0;
      animal._gestationLitterSize = delta._gestationLitterSize || 0;
      if (delta.actionHistory) animal.actionHistory = delta.actionHistory;
      animal._dirty = true;

      if (animal.alive) {
        w.placeAnimal(animal.x, animal.y);
      } else {
        this._deadThisTick.push(animal);
      }
    }

    // Apply plant changes from sub-workers (first come wins)
    const eatenPlants = new Set();
    for (const r of results) {
      if (!r.plantChanges) continue;
      for (const [x, y, ptype, pstage] of r.plantChanges) {
        const idx = y * w.width + x;
        if (eatenPlants.has(idx)) continue;
        eatenPlants.add(idx);
        w.plantType[idx] = ptype;
        w.plantStage[idx] = pstage;
        w.plantAge[idx] = 0;
        if (ptype === 0) w.activePlantTiles.delete(idx);
        w.plantChanges.push([x, y, ptype, pstage]);
      }
    }

    // Add births — reassign proper IDs from the main world counter (with pop cap)
    for (const r of results) {
      if (!r.births) continue;
      for (const bd of r.births) {
        const sc = w.config.animal_species[bd.species];
        if (!sc) continue;
        if (w.isTileOccupied(bd.x, bd.y)) continue;
        // Enforce population cap at merge time
        const bMax = sc.max_population;
        if (bMax) {
          const gMax = w.config.max_animal_population;
          const eMax = gMax > 0 ? Math.max(2, Math.round(bMax * gMax / BASE_POP_TOTAL)) : bMax;
          if (w.getAliveSpeciesCount(bd.species) >= eMax) continue;
        }
        const baby = new Animal(w.nextId(), bd.x, bd.y, bd.species, sc);
        baby.energy = bd.energy;
        baby.sex = bd.sex;
        baby.age = 0;
        baby.pregnant = bd.pregnant || false;
        baby.gestationTimer = bd.gestationTimer || 0;
        baby._gestationLitterSize = bd._gestationLitterSize || 0;
        baby._dirty = true;
        baby._birthTick = w.clock.tick;
        baby.actionHistory = bd.actionHistory || [];
        w.animals.push(baby);
        w.placeAnimal(bd.x, bd.y);
      }
    }

    // Merge egg births from sub-workers (with pop cap including pending eggs)
    if (!w.eggs) w.eggs = [];
    for (const r of results) {
      if (!r.eggBirths) continue;
      for (const ed of r.eggBirths) {
        const sc = w.config.animal_species[ed.species];
        if (!sc) continue;
        // Check pop cap (alive + pending eggs)
        const bMax = sc.max_population;
        if (bMax) {
          const gMax = w.config.max_animal_population;
          const eMax = gMax > 0 ? Math.max(2, Math.round(bMax * gMax / BASE_POP_TOTAL)) : bMax;
          const alive = w.getAliveSpeciesCount(ed.species);
          const pending = w.eggs.filter(e => e.alive && e.species === ed.species).length;
          if ((alive + pending) >= eMax) continue;
        }
        const egg = new Egg(w.nextId(), ed.x, ed.y, ed.species, sc);
        egg.parentA = ed.parentA;
        egg.parentB = ed.parentB;
        egg._birthTick = w.clock.tick;
        w.eggs.push(egg);
      }
    }

    // Rebuild spatial hash from scratch (cheaper than incremental after full merge)
    this.spatialHash.rebuild(w.animals.filter(a => a.alive));

    // Dead animals are already excluded from animalGrid and spatialHash above,
    // so clear _deadThisTick to prevent tickCleanup from double-removing them.
    this._deadThisTick = [];
  }

  /**
   * Process egg incubation, hatching, and cleanup.
   */
  tickEggs() {
    const w = this.world;
    if (!w.eggs || w.eggs.length === 0) return;

    const hatched = [];
    for (const egg of w.eggs) {
      if (!egg.alive) continue;
      egg.tick();
      if (egg.ready) {
        // Hatch: spawn baby animals
        const speciesConfig = w.config.animal_species[egg.species];
        if (!speciesConfig) continue;
        const walkableSet = new Set(speciesConfig.walkable_terrain || [1, 2, 3, 5, 8]);
        const baseTx = egg.x | 0;
        const baseTy = egg.y | 0;
        // Population cap check (with soft cap like behaviors.js)
        const baseMax = speciesConfig.max_population;
        const globalMax = w.config.max_animal_population;
        const effectiveMax = (baseMax && globalMax > 0)
          ? Math.max(2, Math.round(baseMax * globalMax / BASE_POP_TOTAL))
          : (baseMax || Infinity);
        const count = w.getAliveSpeciesCount(egg.species);
        let capBlocked = count >= effectiveMax;
        if (!capBlocked && effectiveMax > 0) {
          const ratio = count / effectiveMax;
          if (ratio > 0.6) {
            const chance = 1 - ((ratio - 0.6) / 0.4);
            if (Math.random() > chance) capBlocked = true;
          }
        }
        if (!capBlocked) {
          // Find a walkable tile for the baby (prefer the egg's own tile)
          let bx = egg.x, by = egg.y;
          let placed = false;
          if (w.isWalkableFor(baseTx, baseTy, walkableSet) && !w.isTileOccupied(baseTx, baseTy)) {
            placed = true;
          } else {
            for (const [ddx, ddy] of [[0, 1], [1, 0], [0, -1], [-1, 0], [1, 1], [-1, 1], [1, -1], [-1, -1]]) {
              const ntx = baseTx + ddx, nty = baseTy + ddy;
              if (w.isWalkableFor(ntx, nty, walkableSet) && !w.isTileOccupied(ntx, nty)) {
                bx = ntx + 0.5; by = nty + 0.5; placed = true; break;
              }
            }
          }
          if (placed) {
            const baby = new Animal(w.nextId(), bx, by, egg.species, speciesConfig);
            baby.energy = speciesConfig.max_energy * 0.4;
            baby.age = 0;
            baby._birthTick = w.clock.tick;
            baby._dirty = true;
            baby.logAction(w.clock.tick, 'HATCHED', { eggId: egg.id, parentA: egg.parentA, parentB: egg.parentB });
            w.animals.push(baby);
            w.placeAnimal(bx, by);
            this.spatialHash.insert(baby);
          }
        }
        hatched.push(egg);
      }
    }
    // Remove hatched and destroyed eggs
    if (hatched.length > 0 || w.eggs.some(e => !e.alive)) {
      w.eggs = w.eggs.filter(e => e.alive && !hatched.includes(e));
    }
  }

  /**
   * Phase 3: Cleanup dead animals and record stats.
   */
  tickCleanup() {
    const w = this.world;
    const profiling = this.profilingEnabled;

    const cleanupStart = performance.now();
    const deadThisTick = this._deadThisTick || [];
    for (const dead of deadThisTick) {
      this.spatialHash.remove(dead);
      w.vacateAnimal(dead.x, dead.y);
    }

    const tick = w.clock.tick;
    const totalAnimals = w.animals.length;
    const cleanupInterval = totalAnimals > 1500 ? 10 : totalAnimals > 800 ? 25 : 50;
    if (tick % cleanupInterval === 0) {
      w.animals = w.animals.filter(a =>
        a.alive || (!a.consumed && a._deathTick != null && tick - a._deathTick < 300)
      );
    }
    this._phases.cleanupMs = performance.now() - cleanupStart;

    // Record stats every 10 ticks
    const statsStart = performance.now();
    if (tick % 10 === 0) {
      const stats = w.getStats();
      stats.tick = tick;
      stats.plant_events = w.plantEvents;
      w.resetPlantEvents();
      w.statsHistory.push(stats);
      if (w.statsHistory.length > 1000) {
        w.statsHistory = w.statsHistory.slice(-1000);
      }
    }
    this._phases.statsMs = performance.now() - statsStart;
    const tickMs = performance.now() - this._tickStart;

    this._latestPhases = this._phases;
    this._latestTickMs = tickMs;

    if (profiling) {
      benchmarkAdd(this._benchmarkCollector, 'animalsProcessed', this._processedAnimals || 0);
      let animalsAlive = 0;
      for (const a of w.animals) {
        if (a.alive) animalsAlive++;
      }
      this._latestProfile = {
        tick: tick,
        tickMs,
        phases: this._phases,
        counts: {
          animalsTotal: w.animals.length,
          animalsAlive,
          activePlants: w.activePlantTiles.size,
        },
      };
    }
  }

  /**
   * Get state for a viewport rectangle.
   */
  getStateForViewport(vx, vy, vw, vh, buffer = 20) {
    const w = this.world;
    if (!w) return null;

    const x1 = Math.max(0, vx - buffer);
    const y1 = Math.max(0, vy - buffer);
    const x2 = Math.min(w.width, vx + vw + buffer);
    const y2 = Math.min(w.height, vy + vh + buffer);

    const animalsData = [];
    for (const a of w.animals) {
      if (a.x >= x1 && a.x < x2 && a.y >= y1 && a.y < y2) {
        if (a.alive || a.state === 9) {
          animalsData.push(a.toDict());
        }
      }
    }

    const eggsData = [];
    for (const egg of (w.eggs || [])) {
      if (egg.alive && egg.x >= x1 && egg.x < x2 && egg.y >= y1 && egg.y < y2) {
        eggsData.push(egg.toDict());
      }
    }

    return {
      clock: w.clock.toDict(),
      animals: animalsData,
      eggs: eggsData,
      plantChanges: w.plantChanges.slice(0, 5000),
    };
  }

  /**
   * Get complete state (for initial load or full sync).
   */
  getFullState() {
    const w = this.world;
    if (!w) return null;

    return {
      clock: w.clock.toDict(),
      animals: w.animals.filter(a => a.alive).map(a => a.toDict()),
      eggs: (w.eggs || []).filter(e => e.alive).map(e => e.toDict()),
      stats: w.getStats(),
    };
  }

  /**
   * Apply terrain edits: array of {x, y, terrain}.
   */
  editTerrain(changes) {
    const w = this.world;
    for (const c of changes) {
      const { x, y, terrain } = c;
      if (w.isInBounds(x, y)) {
        w.terrain[w.idx(x, y)] = terrain;
      }
    }
    // Recompute water proximity
    w.waterProximity = computeWaterProximity(w.terrain, w.width, w.height);
  }

  /**
   * Place a new entity at (x, y). Returns dict or null.
   */
  placeEntity(entityType, x, y) {
    const w = this.world;
    if (!w.isInBounds(x, y)) return null;

    // Animal placement
    const speciesConfig = w.config.animal_species[entityType];
    if (speciesConfig) {
      if (w.isTileOccupied(x, y)) return null;
      const animal = new Animal(w.nextId(), x + 0.5, y + 0.5, entityType, speciesConfig);
      w.animals.push(animal);
      w.placeAnimal(x, y);
      this.spatialHash.insert(animal);
      return animal.toDict();
    }

    // Plant placement
    const TYPE_MAP = {
      APPLE_TREE: P_APPLE_TREE, MANGO_TREE: P_MANGO_TREE,
      STRAWBERRY: P_STRAWBERRY, BLUEBERRY: P_BLUEBERRY,
      GRASS_PLANT: P_GRASS, CARROT: P_CARROT,
      POTATO: P_POTATO, CHILI_PEPPER: P_CHILI_PEPPER,
      OLIVE_TREE: P_OLIVE_TREE,
    };
    if (TYPE_MAP[entityType] !== undefined) {
      const idx = w.idx(x, y);
      w.plantType[idx] = TYPE_MAP[entityType];
      w.plantStage[idx] = S_ADULT;
      w.plantAge[idx] = 100;
      w.activePlantTiles.add(idx);
      return { type: entityType, x, y };
    }

    return null;
  }

  /**
   * Remove an entity by ID.
   */
  removeEntity(entityId) {
    const w = this.world;
    for (const a of w.animals) {
      if (a.id === entityId) {
        a.alive = false;
        a.state = 9; // DEAD
        a._deathTick = this.world.clock.tick;
        this.spatialHash.remove(a);
        return true;
      }
    }
    return false;
  }
}

File: backend/engine/simulation.py
"""Simulation runner — manages the game loop in a background thread."""

import threading
import time
import random

from engine.world import World, WATER, ROCK
from engine.map_generator import generate_terrain
from engine.entities import Animal
from engine.spatial_hash import SpatialHash
from engine.flora import seed_initial_plants, process_plants
from engine.behaviors import decide_and_act


class SimulationRunner:
    """Runs the simulation loop in a background thread."""

    def __init__(self, config, on_tick=None):
        self.config = config
        self.world = None
        self.spatial_hash = SpatialHash(cell_size=16)
        self.on_tick = on_tick  # callback(world) called after each tick

        self._running = False
        self._paused = True
        self._thread = None
        self._tps = config.get("ticks_per_second", 10)
        self._lock = threading.Lock()

        self._step_event = threading.Event()

    def generate_world(self):
        """Generate a new world with terrain, plants, and animals."""
        with self._lock:
            self.world = World(self.config)
            terrain, water_prox, seed = generate_terrain(self.config)
            self.world.terrain = terrain
            self.world.water_proximity = water_prox

            # Seed plants
            seed_initial_plants(self.world)

            # Spawn animals
            self._spawn_animals()

            # Build spatial hash
            self.spatial_hash.rebuild(self.world.animals)

        return seed

    def _spawn_animals(self):
        """Place initial animals on walkable terrain."""
        w = self.world
        h, width = w.terrain.shape

        for species, count_key in [("HERBIVORE", "initial_herbivore_count"),
                                    ("CARNIVORE", "initial_carnivore_count")]:
            count = self.config.get(count_key, 10)
            species_config = self.config["animal_species"][species]
            placed = 0
            attempts = 0
            while placed < count and attempts < count * 50:
                x = random.randint(0, width - 1)
                y = random.randint(0, h - 1)
                if w.terrain[y, x] not in (WATER, ROCK):
                    animal = Animal(w.next_id(), x, y, species, species_config)
                    w.animals.append(animal)
                    placed += 1
                attempts += 1

    def start(self):
        """Start the simulation thread."""
        if self._running:
            return
        if self.world is None:
            self.generate_world()
        self._running = True
        self._paused = False
        self._thread = threading.Thread(target=self._loop, daemon=True)
        self._thread.start()

    def pause(self):
        self._paused = True

    def resume(self):
        self._paused = False

    def step(self):
        """Advance exactly one tick (while paused)."""
        self._step_event.set()

    def stop(self):
        self._running = False
        self._step_event.set()
        if self._thread:
            self._thread.join(timeout=2)
            self._thread = None

    def set_speed(self, tps):
        self._tps = max(1, min(120, tps))

    @property
    def tps(self):
        return self._tps

    @property
    def paused(self):
        return self._paused

    def _loop(self):
        """Main simulation loop."""
        while self._running:
            if self._paused:
                # Wait for step event or unpause
                self._step_event.wait(timeout=0.1)
                if self._step_event.is_set():
                    self._step_event.clear()
                    if self._running:
                        self._tick()
                continue

            start = time.perf_counter()
            self._tick()
            elapsed = time.perf_counter() - start

            # Sleep to maintain target TPS
            target_dt = 1.0 / self._tps
            sleep_time = target_dt - elapsed
            if sleep_time > 0:
                time.sleep(sleep_time)

    def _tick(self):
        """Process one simulation tick."""
        with self._lock:
            w = self.world
            w.plant_changes = []

            # Advance clock
            w.clock.advance()

            # Process flora
            process_plants(w)

            # Process fauna
            alive_animals = [a for a in w.animals if a.alive]
            for animal in alive_animals:
                decide_and_act(animal, w, self.spatial_hash)

            # Update spatial hash for moved/dead animals
            self.spatial_hash.rebuild([a for a in w.animals if a.alive])

            # Clean up dead animals periodically (every 50 ticks)
            if w.clock.tick % 50 == 0:
                w.animals = [a for a in w.animals if a.alive]

            # Record stats periodically
            if w.clock.tick % 10 == 0:
                stats = w.get_stats()
                stats["tick"] = w.clock.tick
                w.stats_history.append(stats)
                # Keep last 1000 samples
                if len(w.stats_history) > 1000:
                    w.stats_history = w.stats_history[-1000:]

        # Notify listeners
        if self.on_tick:
            self.on_tick(w)

    def get_state_for_viewport(self, vx, vy, vw, vh, buffer=20):
        """Get simulation state scoped to a viewport rectangle."""
        with self._lock:
            w = self.world
            if w is None:
                return None

            # Viewport bounds with buffer
            x1 = max(0, vx - buffer)
            y1 = max(0, vy - buffer)
            x2 = min(w.width, vx + vw + buffer)
            y2 = min(w.height, vy + vh + buffer)

            # Animals in viewport
            animals_data = []
            for a in w.animals:
                if a.alive and x1 <= a.x < x2 and y1 <= a.y < y2:
                    animals_data.append(a.to_dict())

            return {
                "clock": w.clock.to_dict(),
                "animals": animals_data,
                "plant_changes": w.plant_changes[:5000],  # cap for bandwidth
            }

    def get_full_state(self):
        """Get complete state for initial load."""
        with self._lock:
            w = self.world
            if w is None:
                return None

            return {
                "clock": w.clock.to_dict(),
                "animals": [a.to_dict() for a in w.animals if a.alive],
                "stats": w.get_stats(),
            }

    def edit_terrain(self, changes):
        """Apply terrain edits: list of {x, y, terrain}."""
        with self._lock:
            w = self.world
            for c in changes:
                x, y, t = c["x"], c["y"], c["terrain"]
                if 0 <= x < w.width and 0 <= y < w.height:
                    w.terrain[y, x] = t

            # Recompute water proximity (full recompute for simplicity)
            from engine.map_generator import _compute_water_proximity
            w.water_proximity = _compute_water_proximity(w.terrain)

    def place_entity(self, entity_type, x, y):
        """Place a new entity at (x, y)."""
        with self._lock:
            w = self.world
            if not (0 <= x < w.width and 0 <= y < w.height):
                return None

            if entity_type in ("HERBIVORE", "CARNIVORE"):
                species_config = w.config["animal_species"][entity_type]
                animal = Animal(w.next_id(), x, y, entity_type, species_config)
                w.animals.append(animal)
                self.spatial_hash.insert(animal)
                return animal.to_dict()
            elif entity_type in ("TREE", "BUSH", "GRASS_PLANT"):
                from engine.flora import P_TREE, P_BUSH, P_GRASS, S_MATURE
                type_map = {"TREE": P_TREE, "BUSH": P_BUSH, "GRASS_PLANT": P_GRASS}
                ptype = type_map[entity_type]
                w.plant_grid[y, x]["type"] = ptype
                w.plant_grid[y, x]["stage"] = S_MATURE
                w.plant_grid[y, x]["age"] = 100
                w.plant_grid[y, x]["fruit"] = 0
                return {"type": entity_type, "x": x, "y": y}
            return None

    def remove_entity(self, entity_id):
        """Remove an entity by ID."""
        with self._lock:
            w = self.world
            for a in w.animals:
                if a.id == entity_id:
                    a.alive = False
                    a.state = 9  # DEAD
                    self.spatial_hash.remove(a)
                    return True
            return False
